flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/library/SimpleCommunityDetection.java

```diff
@@ -0,0 +1,151 @@
+ */
+public class SimpleCommunityDetection implements GraphAlgorithm<Long, Tuple2<Long, Double>, Double> {
+
+	private Integer maxIterations;
+
+	public SimpleCommunityDetection(Integer maxIterations) {
+
+		this.maxIterations = maxIterations;
+	}
+
+	@Override
+	public Graph<Long, Tuple2<Long, Double>, Double> run(Graph<Long, Tuple2<Long, Double>, Double> graph) {
+
+		Graph<Long, Tuple2<Long, Double>, Double> undirectedGraph = graph.getUndirected();
+
+		VertexCentricIteration<Long, Tuple2<Long, Double>, Tuple2<Long, Double>, Double>
+				iteration = undirectedGraph.createVertexCentricIteration(new VertexLabelUpdater(),
+				new LabelMessenger(), maxIterations);
+
+		return undirectedGraph.runVertexCentricIteration(iteration);
+	}
+
+	public static final class VertexLabelUpdater extends VertexUpdateFunction<Long, Tuple2<Long, Double>, Tuple2<Long, Double>> {
+
+		@Override
+		public void updateVertex(Long vertexKey, Tuple2<Long, Double> labelScore,
+								MessageIterator<Tuple2<Long, Double>> inMessages) throws Exception {
+
+			// we would like these two maps to be ordered
+			Map<Long, Double> receivedLabelsWithScores = new TreeMap<Long, Double>();
+			Map<Long, Double> labelsWithHighestScore = new TreeMap<Long, Double>();
+
+			for (Tuple2<Long, Double> message : inMessages) {
+				// split the message into received label and score
+				Long receivedLabel = message.f0;
+				Double receivedScore = message.f1;
+
+				// if the label was received before
+				if (receivedLabelsWithScores.containsKey(receivedLabel)) {
+					Double newScore = receivedScore + receivedLabelsWithScores.get(receivedLabel);
+					receivedLabelsWithScores.put(receivedLabel, newScore);
+				} else {
+					// first time we see the label
+					receivedLabelsWithScores.put(receivedLabel, receivedScore);
+				}
+
+				// store the labels with the highest scores
+				if (labelsWithHighestScore.containsKey(receivedLabel)) {
+					Double currentScore = labelsWithHighestScore.get(receivedLabel);
+					if (currentScore < receivedScore) {
+						// record the highest score
+						labelsWithHighestScore.put(receivedLabel, receivedScore);
+					}
+				} else {
+					// first time we see this label
+					labelsWithHighestScore.put(receivedLabel, receivedScore);
+				}
+			}
+
+			if(receivedLabelsWithScores.size() > 0) {
+				// find the label with the highest score from the ones received
+				Double maxScore = -Double.MAX_VALUE;
+				Long maxScoreLabel = labelScore.f0;
+				for (Long curLabel : receivedLabelsWithScores.keySet()) {
+
+					if (receivedLabelsWithScores.get(curLabel) > maxScore) {
+						maxScore = receivedLabelsWithScores.get(curLabel);
+						maxScoreLabel = curLabel;
+					}
+				}
+
+				// find the highest score of maxScoreLabel
+				Double highestScore = labelsWithHighestScore.get(maxScoreLabel);
+				// re-score the new label
+				if (maxScoreLabel != labelScore.f0) {
+					// delta = 0.5
+					highestScore -= 0.5f / getSuperstepNumber();
```

Comment:
I think delta should be a parameter and the fact that we decrease it by diving with the superstep number should be explained in the description of the algorithm in the beginning.